crypto/ec/ec_lcl.h

```diff
@@ -309,13 +309,11 @@ struct ec_point_st {
 static ossl_inline int ec_point_is_compat(const EC_POINT *point,
                                           const EC_GROUP *group)
 {
-    if (group->meth != point->meth
-        || (group->curve_name != 0
-            && point->curve_name != 0
-            && group->curve_name != point->curve_name))
-        return 0;
-
-    return 1;
+    return (group->meth == point->meth
+            || group->meth->point_cmp == point->meth->point_cmp)
```

Comment:
So the case I have found is the following:
>openssl ecparam -noout -check_named -in ./test/recipes/15-test_
ecparam_data/valid/prime256v1-explicit.pem
It looks like this unnamed curve will get loaded using EC_GFp_mont_method().

The one created form the curve_data is created with this table value...
    {NID_X9_62_prime256v1, &_EC_X9_62_PRIME_256V1.h,
#if defined(ECP_NISTZ256_ASM)
     EC_GFp_nistz256_method,
#elif !defined(OPENSSL_NO_EC_NISTP_64_GCC_128)
     EC_GFp_nistp256_method,
 
So the method tables are not the same (which causes issues with the is_compat)

Reply:
So this happens because EC_GROUP_new_from_ecpkparameters() calls EC_GROUP_new_curve_GFp() with uses either EC_GFp_mont_method() OR EC_GFp_nist_method() WHICH ARE STILL DIFFERENT FROM THE one chosen by ec_group_new_from_data() (which is EC_GFp_nistz256_method).
@romen Do you have a suggestion on the proper way to fix this :)

Reply:
Does there need to be duplicate entries for the different method present in this curve table? Anything in this curve table that sets a custom EC_METHOD will have the same issue I assume.

Reply:
I have changed it back - the real problem is discussed below.

Reply:
This is exactly why I wanted to iterate the tests on all the curves.

We have a few concurrent issues I believe:
- different named curves could share the same set of parameters (aliases)
- a given curve, even if it has no aliases, could have a specialized `EC_METHOD` used when the `EC_GROUP` is created with `EC_GROUP_new_by_curve_name()` or with  `EC_GROUP_new_from_ecpkparameters()`
- `EC_GROUP_cmp()` is tricky:
   ~~~C
    int nid, rv;
    EC_GROUP *group1 = NULL, *group2 = NULL;
    ECPARAMETERS *ecparameters = NULL;
    BN_CTX *ctx = BN_CTX_new();

    group1 = EC_GROUP_new_by_curve_name(nid);
    ecparameters = EC_GROUP_get_ecparameters(group1, NULL);
    group2 = EC_GROUP_new_from_ecparameters(ecparameters);

    rv = EC_GROUP_cmp(group1, group2, NULL);
   
   ~~~
    `rv` is going to be `0` for most curves, but not for all, e.g. the `nistz256` case @slontis mentioned (and in general whenever there is a specialized `EC_METHOD` for a curve).

    It fails to give always the same result because internally `EC_GROUP_cmp(a,b,ctx)` after matching all the other parameters, will do `EC_POINT_cmp(a, Ga, Gb)` where `Ga` and `Gb` are the generator `EC_POINT`s for the two groups, and if `a` and `b` have different `EC_METHOD`s that will fail for `EC_R_INCOMPATIBLE_OBJECTS`.

# Solution A (problematic)

We could fix `EC_GROUP_cmp()` according to its current documentation:

```
=head1 SYNOPSIS

int EC_GROUP_cmp(const EC_GROUP *a, const EC_GROUP *b, BN_CTX *ctx);

=head1 DESCRIPTION

EC_GROUP_cmp compares B<a> and B<b> to determine whether they represent the same curve or not.

=head1 RETURN VALUES

EC_GROUP_cmp returns 0 if the curves are equal, 1 if they are not equal, or -1 on error.
```

My interpretation of its description is that `EC_GROUP_cmp()` should not care about the internal `EC_METHOD` and return equality if two groups _represent_ the same curve (i.e., an `EC_GROUP` derived from `EC_GFp_nistz256_method()` represents the same curve as an `EC_GROUP` derived from `EC_GFp_nistp256_method()` or an `EC_GROUP` derived from `EC_GFp_simple_method`/`EC_GFp_mont_method`/`EC_GFp_nist_method` (with the NIST P-256 parameters).
I could of course be mistaken in my interpretation, but the fact that `EC_GROUP_cmp()` evidently avoids comparing `a->meth` and `b->meth` directly, and that it only happens indirectly through `EC_POINT_cmp()` after checking everything else, suggests to me that at least when it was first implemented the intent was to compare the curves represented by the two groups and not the implementation details.

## One way to fix this

We could replace the `EC_POINT_cmp()` inside `EC_GROUP_cmp()` by taking the affine representation of `gen_a` and `gen_b` and do two `BN_cmp()`; this avoids mixing incompatible points and groups.

## The problem

I tested in `1.0.2a` (from 2015) and `1.1.1b`, and for a long time `EC_GROUP_cmp()` has been throwing `EC_R_INCOMPATIBLE_OBJECTS` from the inner `EC_POINT_cmp()` when the two groups have different `EC_METHOD`s internally.

There was a relevant discussion in #6302 about mixing `EC_POINT P` and `EC_GROUP g` in function calls, if `P` was not created by `EC_POINT_new(g)`.
- Is it a programmer error?
- Do the API signatures suggest that it should be safe (at least as in no `SEGFAULT`s)?

There were different opinions, and that is where `ec_point_is_compat()` was born.
Nonetheless `EC_GROUP_cmp()` is an example of a place inside the library where we can end up calling an `EC_POINT_*()` function with a group that is not the creator for the given point.

@mattcaswell can share more insight about it, but at the end of the day documentation and implementation seem to diverge, and a decision should be taken if it should be fixed adjusting the documentation to the actual behavior or fixing the implementation according to the description (and what to do in 1.0.2, 1.1.0 and 1.1.1) (both changes appear to me as potentially breaking, but probably adding a caveat to the documentation although less desirable is the less disruptive change for existing applications).

# Solution B (subpar)

Avoid using `EC_GROUP_cmp()` to compare groups having potentially different internal implementations but representing the same curve.

I guess we could add an internal `ec_GROUP_cmp_nometh()` that duplicates what `EC_GROUP_cmp()` does but uses `BN_cmp()` on the affine `(x,y)` representations of the groups' generators.

The downside is having to duplicate that code and having to maintain two functions doing basically the same thing in a slightly different way.

# Solution C (maybe the easiest for this PR)

Avoid using `EC_GROUP_cmp()` in `EC_GROUP_check_named_curve()` and instead encode `seed`, `p`, `a`, `b`, `x`, `y`, and `order` from the input `group` through `BN_bn2binpad(*, params_len)` and compare with `memcmp` against the entries in the internal table.

Reply:
For (A) EC_POINT_get_affine_coordinates() also does the is_point_compat() check.

Reply:
> For (A) EC_POINT_get_affine_coordinates() also does the is_point_compat() check.

The idea is to do

~~~C
// BIGNUM *xa, *ya, *xb, *yb;
// {xa,ya,xb,yb} = BN_CTX_get(ctx);

// then instead of EC_POINT_cmp(a, EC_GROUP_get0_generator(a), EC_GROUP_get0_generator(b), ctx)

EC_POINT_get_affine_coordinates(a, EC_GROUP_get0_generator(a), xa, ya, ctx);
EC_POINT_get_affine_coordinates(b, EC_GROUP_get0_generator(b), xb, yb, ctx);

if (r || BN_cmp(xa, xb) != 0 || BN_cmp(ya, yb) != 0)
    r = 1;
~~~

`ec_point_is_compat` is always happy because you never mix methods of group `a` on a `b_point`.

Reply:
Thanks for writing that (it must of taken quite a while).
Maybe C is the right way forward considering the backwards compatibility issues. (NOTE: seed has its own unique size).

Reply:
I tried to implement case C:
There is one repeated failure for the new alias code in ectest.c
    if (rv != nid) {
        /* Fail if the returned nid is not an alias of the original group */
        if (!TEST_ptr(galias = EC_GROUP_new_by_curve_name(rv)))
            goto err;
        EC_GROUP_set_curve_name(galias, nid);

Since trying to get the affine x, y points will have a different curve_name for the group and the point. Which will fail the ec_point_is_compat() which checks the curve_name is the same.
So the curve_name test also breaks things as well as the method test :)

Reply:
Hmmm A random thought..
Maybe setting the API that sets the curve name should be setting the curve name in the point also.

Reply:
I believe that even better would be to have a reference to the creating `EC_GROUP` in the `EC_POINT` structure instead of the name, although we will then need to make sure that no `EC_POINT` outlives an `EC_GROUP`, probably adding refcounting.